Library/OcCryptoLib/RsaDigitalSign.c

```diff
@@ -0,0 +1,511 @@
+};
+
+UINT64
+Mula32 (
+  UINT32  A,
+  UINT32  B,
+  UINT32  C
+  )
+{
+  UINT64 Ret = A;
+
+  Ret *= B;
+  Ret += C;
+  return Ret;
+}
+
+UINT64
+Mulaa32 (
+  UINT32  A,
+  UINT32  B,
+  UINT32  C,
+  UINT32  D
+  )
+{
+  UINT64 Ret = A;
+
+  Ret *= B;
+  Ret += C;
+  Ret += D;
+  return Ret;
+}
+
+//
+//  A[] -= Mod
+//
+STATIC
+VOID
+SubMod (
+  RSA_PUBLIC_KEY  *Key,
+  UINT32          *A
+  )
+{
+  INT64  B     = 0;
+  UINT32 Index = 0;
+  for (Index = 0; Index < Key->Size; Index++) {
+    B += (UINT64) A[Index] - Key->N[Index];
+    A[Index] = (UINT32) B;
+    B >>= 32;
+  }
+}
+
+//
+// Return A[] >= Mod
+//
+STATIC
+INT32
+GeMod (
+  RSA_PUBLIC_KEY  *Key,
+  CONST UINT32    *A
+  )
+{
+  UINT32 Index = 0;
+
+  for (Index = Key->Size; Index;) {
+    Index--;
+    if (A[Index] < Key->N[Index]){
+      return 0;
+    }
+    if (A[Index] > Key->N[Index]){
+      return 1;
+    }
+  }
+  return 1;
+}
+
+//
+// Montgomery c[] += a * b[] / R % mod
+//
+STATIC
+VOID
+MontMulAdd (
+  RSA_PUBLIC_KEY  *Key,
+  UINT32          *C,
+  UINT32          Aa,
+  UINT32          *Bb
+  )
+{
+  UINT64 A, B;
+  UINT32 D0, Index;
+
+  A = Mula32 (Aa, Bb[0], C[0]);
+  D0 = (UINT32) A * Key->N0Inv;
+  B = Mula32 (D0, Key->N[0], (UINT32) A);
+
+  for (Index = 1; Index < Key->Size; Index++) {
+    A = Mulaa32 (Aa, Bb[Index], C[Index], (UINT32) (A >> 32));
+    B = Mulaa32 (D0, Key->N[Index], (UINT32) A, (UINT32) (B >> 32));
+    C[Index - 1] = (UINT32) B;
+  }
+
+  A = (A >> 32) + (B >> 32);
+  C[Index - 1] = (UINT32) A;
+
+  if (A >> 32) {
+    SubMod (Key, C);
+  }
+}
+
+//
+// Montgomery c[] = a[] * b[] / R % mod
+//
+STATIC
+VOID
+MontMul (
+  RSA_PUBLIC_KEY  *Key,
+  UINT32          *C,
+  UINT32          *A,
+  UINT32          *B
+  )
+{
+  UINT32 Index;
+
+  ZeroMem (C, Key->Size);
+
+  for (Index = 0; Index < Key->Size; Index++) {
+    MontMulAdd (Key, C, A[Index], B);
+  }
+}
+
+/**
+  In-place public exponentiation.
+  Exponent depends on the configuration (65537 (default), or 3).
+
+  @param Key        Key to use in signing
+  @param InOut      Input and output big-endian byte array
+ **/
+STATIC
+VOID
+ModPow (
+  RSA_PUBLIC_KEY  *Key,
+  UINT8           *InOut
+  )
+{
+  UINT32 *A     = NULL;
+  UINT32 *Ar    = NULL;
+  UINT32 *Aar   = NULL;
+  UINT32 *Aaa   = NULL;
+  INT32  Index  = 0;
+  UINT32 Tmp    = 0;
+
+  A = AllocateZeroPool (Key->Size * sizeof (UINT32));
+  Ar = AllocateZeroPool (Key->Size * sizeof (UINT32));
+  Aar = AllocateZeroPool (Key->Size * sizeof (UINT32));
+  if (A == NULL || Ar == NULL || Aar == NULL) {
+    if (A != NULL) {
+      FreePool (A);
+    }
+    if (Ar != NULL) {
+      FreePool (Ar);
+    }
+    if (Aar != NULL) {
+      FreePool (Aar);
+      // TODO: debug_error about efi memory alocation
+    }
+    return;
+  }
+
+  //
+  // Re-use location
+  //
+  Aaa = Aar;
+
+  //
+  // Convert from big endian byte array to little endian word array
+  //
+  for (Index = 0; Index < (INT32) Key->Size; Index++) {
```

Comment:
Why Index is signed here?

Reply:
Because just a bit below we have this loop: 
```
for (Index = (INT32) Key->Size - 1; Index >= 0; --Index) {
    Tmp = Aaa[Index];

    *InOut++ = (UINT8) (Tmp >> 24);
    *InOut++ = (UINT8) (Tmp >> 16);
    *InOut++ = (UINT8) (Tmp >>  8);
    *InOut++ = (UINT8) (Tmp >>  0);
  }
```
If we use unsigned int for index, then this loop will become endless.

Reply:
Just turn this loop into a do while, and it will work nicely. Rewrote with improved compiler compatibility.
```C
  Index = Key->Size;
  do {
    Tmp = Aaa[--Index];
    *InOut++ = (UINT8) ((Tmp >> 24U) & 0xFFU);
    *InOut++ = (UINT8) ((Tmp >> 16U) & 0xFFU);
    *InOut++ = (UINT8) ((Tmp >>  8U) & 0xFFU);
    *InOut++ = (UINT8) ((Tmp >>  0U) & 0xFFU);
  } while (Index > 0);
```

Reply:
Done. https://github.com/acidanthera/OcSupportPkg/commit/53bee39c816b29514d8bdb17369ee6d2906c2c92